src/Npgsql/UnpooledConnectorSource.cs

```diff
@@ -20,10 +22,20 @@ public UnpooledConnectorSource(NpgsqlConnectionStringBuilder settings, string co
         internal override async ValueTask<NpgsqlConnector> Get(
             NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
         {
-            var connector = new NpgsqlConnector(conn, this);
-            await connector.Open(timeout, async, cancellationToken);
-            Interlocked.Increment(ref _numConnectors);
-            return connector;
+            try
+            {
+                var connector = new NpgsqlConnector(this, conn);
+                await connector.Open(timeout, async, cancellationToken);
+                connector.Connection = conn;
```

Comment:
Should we remove the mutual setting of references between connection/connector, and do it outside? Now that these two types are more cleanly separate, it seems like the connector sources should really only return a connector, and nothing more. The only reason we still need a connection is because the connection has some callbacks necessary for establishing the connection.